source/subscription-manager/subscription_manager.c

```diff
@@ -0,0 +1,130 @@
+                ( pxSubscriptionList[ lIndex ].pvIncomingPublishCallbackContext == pvIncomingPublishCallbackContext ) )
+            {
+                LogWarn( ( "Subscription already exists.\n" ) );
+                xAvailableIndex = SUBSCRIPTION_MANAGER_MAX_SUBSCRIPTIONS;
+                xReturnStatus = pdTRUE;
+                break;
+            }
+        }
+    }
+
+    if( ( xAvailableIndex < SUBSCRIPTION_MANAGER_MAX_SUBSCRIPTIONS ) && ( pxIncomingPublishCallback != NULL ) )
+    {
+        pxSubscriptionList[ xAvailableIndex ].pcSubscriptionFilterString = pcTopicFilterString;
+        pxSubscriptionList[ xAvailableIndex ].uFilterStringLength = uTopicFilterLength;
+        pxSubscriptionList[ xAvailableIndex ].pxIncomingPublishCallback = pxIncomingPublishCallback;
+        pxSubscriptionList[ xAvailableIndex ].pvIncomingPublishCallbackContext = pvIncomingPublishCallbackContext;
+        xReturnStatus = pdTRUE;
+    }
+
+    return xReturnStatus;
+}
+
+/*-----------------------------------------------------------*/
+
+void removeSubscription( SubscriptionElement_t * pxSubscriptionList,
+                         const char * topicFilterString,
+                         uint16_t topicFilterLength )
+{
+    int32_t lIndex = 0;
+
+    for( lIndex = 0; lIndex < SUBSCRIPTION_MANAGER_MAX_SUBSCRIPTIONS; lIndex++ )
+    {
+        if( pxSubscriptionList[ lIndex ].uFilterStringLength == topicFilterLength )
+        {
+            if( strncmp( pxSubscriptionList[ lIndex ].pcSubscriptionFilterString, topicFilterString, topicFilterLength ) == 0 )
+            {
+                memset( &( pxSubscriptionList[ lIndex ] ), 0x00, sizeof( SubscriptionElement_t ) );
+            }
+        }
+    }
+}
+
+/*-----------------------------------------------------------*/
+
+BaseType_t handleIncomingPublishes( SubscriptionElement_t * pxSubscriptionList,
+                                    MQTTPublishInfo_t * pxPublishInfo )
+{
+    BaseType_t xCallbackInvoked = pdFALSE;
+    int32_t lIndex = 0;
+    bool isMatched = false;
+
+    for( lIndex = 0; lIndex < SUBSCRIPTION_MANAGER_MAX_SUBSCRIPTIONS; lIndex++ )
+    {
+        if( pxSubscriptionList[ lIndex ].uFilterStringLength > 0 )
+        {
+            MQTT_MatchTopic( pxPublishInfo->pTopicName,
+                             pxPublishInfo->topicNameLength,
+                             pxSubscriptionList[ lIndex ].pcSubscriptionFilterString,
+                             pxSubscriptionList[ lIndex ].uFilterStringLength,
+                             &isMatched );
+
+            if( isMatched == true )
+            {
+                pxSubscriptionList[ lIndex ].pxIncomingPublishCallback( pxSubscriptionList[ lIndex ].pvIncomingPublishCallbackContext,
+                                                                        pxPublishInfo );
+                xCallbackInvoked = pdTRUE;
+            }
+        }
+    }
+
+    return xCallbackInvoked;
```

Comment:
You could save two lines since `isMatched` and `xCallbackInvoked` will always be identical
```suggestion
    return ( BaseType_t ) isMatched;
```
or
```suggestion
    return ( isMatched ) ? pdTRUE : pdFALSE;
```